### PR TITLE
Docker Clustering: Design proposal.

### PR DESCRIPTION
**Pull request by aluzzardi**
Authors: @aluzzardi and @vieux.

[ScreenShot](https://www.youtube.com/watch?v=vtnSL79rZ6o)

The goal and scope of this change is to allow docker to manage a cluster
of docker hosts.

The target audience for clustering is modeled after Docker itself:
developer and devops first then enterprise later. Enterprise requires
additional features such as authentication, ACLs, auditing and tooling
which will come at a later time.

The system is designed to handle workloads in the same way as Docker: it
can run both long-running and one-off tasks. Batch processing can be
built on top of the API using the one-off primitive.

The architecture is based on an evented model where the master queries
the registered slaves within the system and checks the current state of
the slaves against the requested state by the user.  It reconciles any
differences and updates the cluster state with the required changes.

Signed-off-by: Andrea Luzzardi aluzzardi@gmail.com
Signed-off-by: Victor Vieux vieux@docker.com

### CODE REVIEW

**Comment by boonkerz** [comment]
When container are rebalanced as example an typical webapp
docker run -d --name mysql mysql
docker run -d --name elasticsearch elasticsearch
docker run -d --name web --link mysql:mysql --link elasticsearch:elasticsearch webserver
and the elasticsearch goes down
do the cluster also reconfigure the link on the webserver container?
and when the mysql is started on host 1 and the webserver on host2 
do the cluster connect the right hosts together?

**Comment by vieux** [comment]
@boonkerz right now swarm don't support links.
There is a huge effort to improving the networking model directly into the docker engine.
With this improvements, the links will work between 2 hosts.
We are waiting for this to use it.

**Comment by vieux** [comment]
Hi everyone, 

As you probably figured out, this design proposal was an early version of [Swarm](http://github.com/docker/swarm)

Please redirect all your concerns to the [Swarm issue tracker](https://github.com/docker/swarm/issues)

Thanks!

File: docs/sources/userguide/cluster.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+
+In order to ensure consistency within the cluster one of your Docker Engines
+will need to be promoted to master within the cluster.  If you are using
+the Hub's discovery service you will be able to promote any of your registered
+nodes to a master from the web interface.  You can also statically assign one
+of the Docker Engines with the `--master` flag.
+
+    $ docker -d --master --discovery https://discovery.hub.docker.com/u/demo-user/cluster-1
+
+Eventually, clustering will provide a built-in leader election algorithm making the `--master` flag obsolete.
```
Comment (BhargavaRamM) [inline docs/sources/userguide/cluster.md]: What are the available options for allowing fault tolerance in cluster ?(If master goes down.. one of the nodes has to be picked and given privileges). I want to use Docker swarm orchestration technology and docker swarm doesn't provide any fault tolerance.

**Pull request closed without merging**